File: android/src/main/java/org/dwbn/drmkit/DrmIdentifiers.kt
package org.dwbn.drmkit

import java.util.Base64
import java.util.regex.Pattern
import javax.crypto.Cipher
import javax.crypto.spec.IvParameterSpec
import javax.crypto.spec.SecretKeySpec

object DrmIdentifiers {
    private val UUID_PATTERN: Pattern = Pattern.compile(
        "^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$"
    )

    @JvmStatic
    fun toHex(bytes: ByteArray): String =
        bytes.joinToString("") { "%02x".format(it) }

    @JvmStatic
    fun hexToBytes(hex: String): ByteArray {
        val normalized = hex.lowercase()
        if (normalized.length != 32 || !normalized.all { it in '0'..'9' || it in 'a'..'f' }) {
            throw IllegalArgumentException("Expected 16-byte hex")
        }
        return ByteArray(16) { index ->
            normalized.substring(index * 2, index * 2 + 2).toInt(16).toByte()
        }
    }

    @JvmStatic
    fun uuidToBytes(uuid: String): ByteArray = hexToBytes(uuid.replace("-", ""))

    @JvmStatic
    fun fairPlayUri(kidHex: String, ivHex: String): String =
        "skd://${kidHex.lowercase()}:${ivHex.lowercase()}"

    /**
     * RFC 4122 UUID from KID bytes (big-endian, lowercase, dashed). Not mixed-endian.
     */
    @JvmStatic
    fun axinomKeyId(kidBytes: ByteArray): String {
        val hex = toHex(kidBytes)
        return "${hex.substring(0, 8)}-${hex.substring(8, 12)}-${hex.substring(12, 16)}-${hex.substring(16, 20)}-${hex.substring(20, 32)}"
    }

    @JvmStatic
    fun axinomKeyValue(keyBytes: ByteArray): String =
        Base64.getEncoder().encodeToString(keyBytes)

    /**
     * AES-128-CBC without padding: first 16 bytes of the communication key, IV = KID bytes.
     */
    @JvmStatic
    fun encryptContentKeyForAxinom(
        contentKey: ByteArray,
        communicationKey: ByteArray,
        kidBytes: ByteArray,
    ): ByteArray {
        val aesKey = communicationKey.copyOfRange(0, minOf(16, communicationKey.size))
        if (contentKey.size != 16 || aesKey.size != 16 || kidBytes.size != 16) {
            throw IllegalArgumentException("Expected 16-byte AES inputs")
        }
        return cryptAes128Cbc(Cipher.ENCRYPT_MODE, contentKey, aesKey, kidBytes)
    }

    @JvmStatic
    fun communicationKeyBytes(value: String): ByteArray {
        val trimmed = value.trim()
        if (UUID_PATTERN.matcher(trimmed).matches()) {
            return uuidToBytes(trimmed)
        }
        val raw = trimmed.toByteArray(Charsets.ISO_8859_1)
        if (raw.size == 16) {
            return raw
        }
        val decoded = try {
            Base64.getDecoder().decode(trimmed)
        } catch (_: IllegalArgumentException) {
            null
        }
        if (decoded != null && (decoded.size == 16 || decoded.size == 32)) {
            return decoded
        }
        throw IllegalArgumentException("Invalid communication key")
    }

    @JvmStatic
    fun communicationKeyBytes(value: ByteArray): ByteArray {
        if (value.size != 16) {
            throw IllegalArgumentException("Invalid communication key")
        }
        return value
    }

    @JvmStatic
    fun additionalData(kidBytes: ByteArray, packageUuidBytes: ByteArray): ByteArray =
        kidBytes + packageUuidBytes

    internal fun decryptContentKeyForAxinom(
        encrypted: ByteArray,
        communicationKey: ByteArray,
        kidBytes: ByteArray,
    ): ByteArray {
        val aesKey = communicationKey.copyOfRange(0, minOf(16, communicationKey.size))
        if (encrypted.size != 16 || aesKey.size != 16 || kidBytes.size != 16) {
            throw IllegalArgumentException("Expected 16-byte AES inputs")
        }
        return cryptAes128Cbc(Cipher.DECRYPT_MODE, encrypted, aesKey, kidBytes)
    }

    private fun cryptAes128Cbc(
        mode: Int,
        data: ByteArray,
        aesKey: ByteArray,
        kidBytes: ByteArray,
    ): ByteArray {
        val cipher = Cipher.getInstance("AES/CBC/NoPadding")
        cipher.init(mode, SecretKeySpec(aesKey, "AES"), IvParameterSpec(kidBytes))
        return cipher.doFinal(data)
    }
}

File: ios/Sources/DrmKit/DrmIdentifiers.swift
import CommonCrypto
import Foundation

public enum DrmIdentifiersError: Error, Equatable {
    case invalidArgument(String)
    case encryptionFailed
}

public enum DrmIdentifiers {
    public static func toHex(_ bytes: Data) -> String {
        bytes.map { String(format: "%02x", $0) }.joined()
    }

    public static func hexToBytes(_ hex: String) throws -> Data {
        let normalized = hex.lowercased()
        guard normalized.count == 32, normalized.allSatisfy({ $0.isHexDigit && $0.isASCII }) else {
            throw DrmIdentifiersError.invalidArgument("Expected 16-byte hex")
        }
        var bytes = Data(capacity: 16)
        var index = normalized.startIndex
        while index < normalized.endIndex {
            let next = normalized.index(index, offsetBy: 2)
            let byteString = normalized[index..<next]
            guard let byte = UInt8(byteString, radix: 16) else {
                throw DrmIdentifiersError.invalidArgument("Expected 16-byte hex")
            }
            bytes.append(byte)
            index = next
        }
        return bytes
    }

    public static func uuidToBytes(_ uuid: String) throws -> Data {
        try hexToBytes(uuid.replacingOccurrences(of: "-", with: ""))
    }

    public static func fairPlayUri(kidHex: String, ivHex: String) -> String {
        "skd://\(kidHex.lowercased()):\(ivHex.lowercased())"
    }

    /// RFC 4122 UUID from KID bytes (big-endian, lowercase, dashed). Not mixed-endian.
    public static func axinomKeyId(_ kidBytes: Data) -> String {
        let hex = toHex(kidBytes)
        let start = hex.startIndex
        func slice(_ offset: Int, _ count: Int) -> Substring {
            let from = hex.index(start, offsetBy: offset)
            let to = hex.index(from, offsetBy: count)
            return hex[from..<to]
        }
        return "\(slice(0, 8))-\(slice(8, 4))-\(slice(12, 4))-\(slice(16, 4))-\(slice(20, 12))"
    }

    public static func axinomKeyValue(_ keyBytes: Data) -> String {
        keyBytes.base64EncodedString()
    }

    /// AES-CBC without padding, IV = KID bytes. 32-byte communication key uses AES-256; 16-byte uses AES-128.
    public static func encryptContentKeyForAxinom(
        contentKey: Data,
        communicationKey: Data,
        kidBytes: Data
    ) throws -> Data {
        guard contentKey.count == 16, kidBytes.count == 16,
              communicationKey.count == 16 || communicationKey.count == 32 else {
            throw DrmIdentifiersError.invalidArgument("Expected a 16-byte content key and a 16- or 32-byte communication key")
        }
        return try cryptAESCBC(
            data: contentKey,
            key: communicationKey,
            iv: kidBytes,
            operation: CCOperation(kCCEncrypt)
        )
    }

    public static func communicationKeyBytes(_ value: String) throws -> Data {
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        let uuidRange = NSRange(location: 0, length: trimmed.utf16.count)
        let uuidPattern = "^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$"
        if let regex = try? NSRegularExpression(pattern: uuidPattern),
           regex.firstMatch(in: trimmed, range: uuidRange) != nil {
            return try uuidToBytes(trimmed)
        }
        let raw = Data(trimmed.utf8)
        if raw.count == 16 {
            return raw
        }
        if let decoded = Data(base64Encoded: trimmed), decoded.count == 16 || decoded.count == 32 {
            return decoded
        }
        throw DrmIdentifiersError.invalidArgument("Invalid communication key")
    }

    public static func communicationKeyBytes(_ value: Data) throws -> Data {
        guard value.count == 16 else {
            throw DrmIdentifiersError.invalidArgument("Invalid communication key")
        }
        return value
    }

    public static func additionalData(kidBytes: Data, packageUuidBytes: Data) -> Data {
        kidBytes + packageUuidBytes
    }

    static func decryptContentKeyForAxinom(
        encrypted: Data,
        communicationKey: Data,
        kidBytes: Data
    ) throws -> Data {
        guard encrypted.count == 16, kidBytes.count == 16,
              communicationKey.count == 16 || communicationKey.count == 32 else {
            throw DrmIdentifiersError.invalidArgument("Expected a 16-byte ciphertext and a 16- or 32-byte communication key")
        }
        return try cryptAESCBC(
            data: encrypted,
            key: communicationKey,
            iv: kidBytes,
            operation: CCOperation(kCCDecrypt)
        )
    }

    private static func cryptAESCBC(
        data: Data,
        key: Data,
        iv: Data,
        operation: CCOperation
    ) throws -> Data {
        let keySize: size_t
        switch key.count {
        case kCCKeySizeAES128:
            keySize = size_t(kCCKeySizeAES128)
        case kCCKeySizeAES256:
            keySize = size_t(kCCKeySizeAES256)
        default:
            throw DrmIdentifiersError.invalidArgument("Expected a 16- or 32-byte AES key")
        }
        var output = Data(count: data.count)
        var outputLength: size_t = 0
        let status = data.withUnsafeBytes { dataBytes in
            key.withUnsafeBytes { keyBytes in
                iv.withUnsafeBytes { ivBytes in
                    output.withUnsafeMutableBytes { outputBytes in
                        CCCrypt(
                            operation,
                            CCAlgorithm(kCCAlgorithmAES),
                            CCOptions(0),
                            keyBytes.baseAddress,
                            keySize,
                            ivBytes.baseAddress,
                            dataBytes.baseAddress,
                            data.count,
                            outputBytes.baseAddress,
                            data.count,
                            &outputLength
                        )
                    }
                }
            }
        }
        guard status == kCCSuccess, outputLength == data.count else {
            throw DrmIdentifiersError.encryptionFailed
        }
        return output
    }
}
